Modules/_decimal/_decimal.c

```diff
@@ -3675,34 +3672,54 @@ dec_as_long(PyObject *dec, PyObject *context, int round)
     }
 
     status = 0;
-    ob_digit = NULL;
-#if PYLONG_BITS_IN_DIGIT == 30
-    n = mpd_qexport_u32(&ob_digit, 0, PyLong_BASE, x, &status);
-#elif PYLONG_BITS_IN_DIGIT == 15
-    n = mpd_qexport_u16(&ob_digit, 0, PyLong_BASE, x, &status);
-#else
-    #error "PYLONG_BITS_IN_DIGIT should be 15 or 30"
-#endif
+    int64_t val = mpd_qget_i64(x, &status);
 
-    if (n == SIZE_MAX) {
-        PyErr_NoMemory();
+    if (!status) {
+        mpd_del(x);
+        return PyLong_FromInt64(val);
+    }
+    assert(!mpd_iszero(x));
+
+    const PyLongLayout *layout = PyLong_GetNativeLayout();
+    const uint32_t base = (uint32_t)1 << layout->bits_per_digit;
+
+    assert(layout->bits_per_digit <= 32);
+    assert(layout->digits_order == -1);
+    assert(layout->digit_endianness == (PY_LITTLE_ENDIAN ? -1 : 1));
+    assert(layout->digit_size == 2 || layout->digit_size == 4);
+
+    size_t n, len = mpd_sizeinbase(x, base);
+    void *digits;
+    PyLongWriter *writer = PyLongWriter_Create(mpd_isnegative(x), len, &digits);
+
+    if (writer == NULL) {
         mpd_del(x);
         return NULL;
     }
 
-    if (n == 1) {
-        sdigit val = mpd_arith_sign(x) * ob_digit[0];
-        mpd_free(ob_digit);
+    status = 0;
+    /* mpd_qexport_*() functions used here with assumption, that no resizing
+       occur, i.e. len was obtained by a call to mpd_sizeinbase.  Set digits
+       to zero, as size can be overestimated. */
+    if (layout->digit_size == 4) {
+        memset(digits, 0, len*sizeof(uint32_t));
```

Comment:
Use either 4 or `sizeof(uint32_t)` in both cases.

You can also only clear uninitialized digits between `n` and `len`, but this will complicate the code. I am not sure the time difference will be noticeable.

Reply:
I will be noticeable in a loop, but for simple benchmarks proposed here - no.